# 4.7: Deriving New Variables

In [4]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [5]:
# Creating a path
path = r'C:\Users\zwier\anaconda_projects\15-01-2025 Instacart Basket Analysis'

In [6]:
# importing df
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

## Creating Subset with a limit

In [8]:
# creating a limited subset
df = ords_prods_merge[:1000000] # ":1000000" means that "limit from beggining to row 1000000"

# If-Statements with User-Defined Functions

## Creating a IF (Case when) 

In [11]:
# Setting a IF statement with 4 options

def price_label(row): # head part of the function

  if row['prices'] <= 5:                                            # Body part od the function
    return 'Low-range product' # results <= 5
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product' # range from results > 5 to <= 15  
  elif row['prices'] > 15:
    return 'High range' # results > 15
  else: return 'Not enough data' # if nan

In [ ]:
  if row['prices'] <= 5:
    return 'Low-range product' # results <= 5
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

## Adding Column to DF APPLY

In [13]:
# lets add Price_label column to the subset
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\zwier\AppData\Local\Temp\ipykernel_20568\3006559838.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [14]:
# Result Check
df.head()

order_id   user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    1107.0   38259.0           2.0                 1.0               11.0   
1    5319.0  196224.0          65.0                 1.0               14.0   
2    7540.0  138499.0           8.0                 0.0               14.0   
3    9228.0   79603.0           2.0                 2.0               10.0   
4    9273.0   50005.0           1.0                 1.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         1.0                7.0        0.0   
1                     1.0         1.0                3.0        1.0   
2                     7.0         1.0                4.0        1.0   
3                    30.0         1.0                2.0        0.0   
4                     0.0         1.0               30.0        0.0   

                 product_name  aisle_id  department_id  prices _merge  \
0  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
1  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
2  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
3  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
4  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   

         price_range  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

In [15]:
# Value check
df['price_range'].value_counts()

price_range
Mid-range product    634889
Low-range product    331220
Not enough data       24547
High range             9344
Name: count, dtype: int64

In [16]:
# dropping '_merge' column = Not needed anymore
# df = df.drop(columns = ['_merge'])

In [17]:
# Shape check
df.shape

(1000000, 15)

In [18]:
# Checking prices
df['prices'].value_counts()

prices
11.3    53649
9.0     44266
3.8     32305
1.4     31093
4.9     30112
        ...  
24.5       17
19.9       11
16.9       11
18.8        6
18.6        6
Name: count, Length: 157, dtype: int64

In [19]:
# Checking Max
df['prices'].max()

24.5

# If-Statements with the loc() Function

In [21]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\zwier\AppData\Local\Temp\ipykernel_20568\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [22]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [23]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [24]:
# Value check
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     634889
Low-range product     331220
NaN                    24547
High-range product      9344
Name: count, dtype: int64

# If-Statements with For-Loops

In [82]:
# Exampe of a Loop

# for x in range(30, 45):
 #    print("My age is %d" % (x))

In [34]:
# checking what Day most orders take place
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0.0    6252941
1.0    5696724
6.0    4530304
2.0    4243173
5.0    4236989
3.0    3868747
4.0    3812390
NaN         11
Name: count, dtype: int64

In [35]:
# For Loop with If

result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [36]:
# checking the new list
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Regularly busy

In [37]:
# Adding result list to a dataframe
ords_prods_merge['busiest_day'] = result

In [38]:
# Checking results
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22575948
Busiest day        6252941
Least busy         3812390
Name: count, dtype: int64

In [39]:
# Checking shape
ords_prods_merge.shape

(32641279, 15)

# Exercise 4.7

### Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [42]:
# For Loop with If
results = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value in [0, 1]:  
        results.append("Busiest days")
    elif value in [4, 3]:  
        results.append("Slowest days")
    else:
        results.append("Regularly busy")

In [43]:
# Adding results list to the dataframe
ords_prods_merge['Busiest days'] = results

In [44]:
# Checking results
ords_prods_merge.head()

order_id   user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    1107.0   38259.0           2.0                 1.0               11.0   
1    5319.0  196224.0          65.0                 1.0               14.0   
2    7540.0  138499.0           8.0                 0.0               14.0   
3    9228.0   79603.0           2.0                 2.0               10.0   
4    9273.0   50005.0           1.0                 1.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         1.0                7.0        0.0   
1                     1.0         1.0                3.0        1.0   
2                     7.0         1.0                4.0        1.0   
3                    30.0         1.0                2.0        0.0   
4                     0.0         1.0               30.0        0.0   

                 product_name  aisle_id  department_id  prices _merge  \
0  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
1  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
2  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
3  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
4  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   

      busiest_day    Busiest days  
0  Regularly busy    Busiest days  
1  Regularly busy    Busiest days  
2     Busiest day    Busiest days  
3  Regularly busy  Regularly busy  
4  Regularly busy    Busiest days

### Check the values of this new column for accuracy. Note any observations in markdown format.

In [51]:
# Checking Values
ords_prods_merge['Busiest days'].value_counts(dropna =False)

Busiest days
Regularly busy    13010477
Busiest days      11949665
Slowest days       7681137
Name: count, dtype: int64

### The Busiest days column shows that most days are categorized as "Regularly busy" (13,010,477), followed by "Busiest days" (11,949,665) and "Slowest days" (7,681,137). The absence of NaN values confirms that all rows have been accurately assigned a category, ensuring data integrity.

In [53]:
# Check
ords_prods_merge.head()

order_id   user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    1107.0   38259.0           2.0                 1.0               11.0   
1    5319.0  196224.0          65.0                 1.0               14.0   
2    7540.0  138499.0           8.0                 0.0               14.0   
3    9228.0   79603.0           2.0                 2.0               10.0   
4    9273.0   50005.0           1.0                 1.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         1.0                7.0        0.0   
1                     1.0         1.0                3.0        1.0   
2                     7.0         1.0                4.0        1.0   
3                    30.0         1.0                2.0        0.0   
4                     0.0         1.0               30.0        0.0   

                 product_name  aisle_id  department_id  prices _merge  \
0  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
1  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
2  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
3  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
4  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   

      busiest_day    Busiest days  
0  Regularly busy    Busiest days  
1  Regularly busy    Busiest days  
2     Busiest day    Busiest days  
3  Regularly busy  Regularly busy  
4  Regularly busy    Busiest days

In [54]:
# Cleaning
# ords_prods_merge = ords_prods_merge.drop(columns = ['busiest_day'])

### When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [56]:
# Value check
ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10.0    2781009
11.0    2755369
14.0    2709084
15.0    2682090
13.0    2680171
12.0    2637533
16.0    2554110
9.0     2471067
17.0    2104225
8.0     1729871
18.0    1649870
19.0    1268365
20.0     983232
7.0      897163
21.0     801052
22.0     638671
23.0     405051
6.0      292533
0.0      220337
1.0      116589
5.0       88661
2.0       69904
4.0       53636
3.0       51675
Name: count, dtype: int64

In [57]:
# Creating If statement

def categorize_hour(hour):
    if 10 <= hour <= 16:
        return "Most orders"
    elif 7 <= hour <= 9 or 17 <= hour <= 20:
        return "Average orders"
    else:
        return "Fewest orders"

In [58]:
# Adding if to a DataFrame
ords_prods_merge['busiest_period_of_day'] = ords_prods_merge['order_hour_of_day'].apply(categorize_hour)

In [69]:
# Checking results
ords_prods_merge.head()

order_id   user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    1107.0   38259.0           2.0                 1.0               11.0   
1    5319.0  196224.0          65.0                 1.0               14.0   
2    7540.0  138499.0           8.0                 0.0               14.0   
3    9228.0   79603.0           2.0                 2.0               10.0   
4    9273.0   50005.0           1.0                 1.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         1.0                7.0        0.0   
1                     1.0         1.0                3.0        1.0   
2                     7.0         1.0                4.0        1.0   
3                    30.0         1.0                2.0        0.0   
4                     0.0         1.0               30.0        0.0   

                 product_name  aisle_id  department_id  prices _merge  \
0  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
1  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
2  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
3  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   
4  Chocolate Sandwich Cookies      61.0           19.0     5.8   both   

      busiest_day    Busiest days busiest_period_of_day  
0  Regularly busy    Busiest days           Most orders  
1  Regularly busy    Busiest days           Most orders  
2     Busiest day    Busiest days           Most orders  
3  Regularly busy  Regularly busy           Most orders  
4  Regularly busy    Busiest days           Most orders

In [60]:
# Checking results
ords_prods_merge[['order_hour_of_day', 'busiest_period_of_day']].value_counts()

order_hour_of_day  busiest_period_of_day
10.0               Most orders              2781009
11.0               Most orders              2755369
14.0               Most orders              2709084
15.0               Most orders              2682090
13.0               Most orders              2680171
12.0               Most orders              2637533
16.0               Most orders              2554110
9.0                Average orders           2471067
17.0               Average orders           2104225
8.0                Average orders           1729871
18.0               Average orders           1649870
19.0               Average orders           1268365
20.0               Average orders            983232
7.0                Average orders            897163
21.0               Fewest orders             801052
22.0               Fewest orders             638671
23.0               Fewest orders             405051
6.0                Fewest orders             292533
0.0                Fewe

### Print the frequency for this new column.

In [76]:
# Value check
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       18799366
Average orders    11103793
Fewest orders      2738120
Name: count, dtype: int64

In [84]:
# Exporting
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge4.7.pkl'))